### PR TITLE
Fix Writer history behavior when there are no matched readers [9030]

### PR DESCRIPTION
**Pull request by laura-eprosima**
* Add Blackbox test 
* Fix bug [8945]

### CODE REVIEW

File: src/cpp/rtps/writer/StatefulWriter.cpp [hunk 1]
```diff
@@ -1455,6 +1452,7 @@ bool StatefulWriter::matched_reader_remove(
         return true;
     }
 
+    check_acked_status();
```
Comment (MiguelCompany) [inline src/cpp/rtps/writer/StatefulWriter.cpp:1455]: As we have not modified the list of matched readers, I don't think this is necessary

File: test/blackbox/common/BlackboxTestsPubSubHistory.cpp [hunk 2]
```diff
@@ -664,6 +664,69 @@ TEST_P(PubSubHistory, ReliableTransientLocalKeepLast1Data300Kb)
     ASSERT_EQ(reader.get_last_sequence_received().low, 10u);
 }
 
+// Test created to check bug #8945
```
Comment (MiguelCompany) [inline test/blackbox/common/BlackboxTestsPubSubHistory.cpp:667]: Should we really use 1600 samples for this? Is it reproducible with a lower number, like 16 ?

File: test/blackbox/common/BlackboxTestsPubSubHistory.cpp [hunk 3]
```diff
@@ -664,6 +664,69 @@ TEST_P(PubSubHistory, ReliableTransientLocalKeepLast1Data300Kb)
...
+/*!
+ * @fn TEST(PubSubHistory, WriterWithoutReadersTransientLocal)
+ * @brief This test checks that the writer doesn't fill its history while there are no readers matched.
+ *
+ * The test creates a Reliable, Transient Local Writer with a Keep All history, and its resources limited to
+ * 1300 samples.
```
Comment (MiguelCompany) [inline test/blackbox/common/BlackboxTestsPubSubHistory.cpp:673]:
```suggestion
 * 13 samples.
```

File: src/cpp/rtps/writer/StatefulWriter.cpp [hunk 4]
```diff
@@ -1657,14 +1656,7 @@ bool StatefulWriter::try_remove_change(
...
         std::lock_guard<RecursiveTimedMutex> guard(mp_mutex);
-        for (ReaderProxy* it : matched_readers_)
-        {
-            SequenceNumber_t reader_low_mark = it->changes_low_mark();
-            if (min_low_mark == SequenceNumber_t() || reader_low_mark < min_low_mark)
-            {
-                min_low_mark = reader_low_mark;
-            }
-        }
+        min_low_mark = next_all_acked_notify_sequence_;
```
Comment (MiguelCompany) [inline src/cpp/rtps/writer/StatefulWriter.cpp:1659]: We should remove 1 in order to be consistent
```suggestion
        min_low_mark = next_all_acked_notify_sequence_ - 1;
```

File: src/cpp/rtps/writer/StatefulWriter.cpp [hunk 5]
```diff
@@ -356,10 +356,7 @@ void StatefulWriter::unsent_change_added_to_history(
                     periodic_hb_event_->restart_timer(max_blocking_time);
                 }
 
-                if ( (mp_listener != nullptr) && this->is_acked_by_all(change) )
-                {
-                    mp_listener->onWriterChangeReceivedByAll(this, change);
-                }
+                check_acked_status();
```
Comment (MiguelCompany) [inline src/cpp/rtps/writer/StatefulWriter.cpp:359]: I think this should be called after the if below

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2484/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2298/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2489/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2563/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2373/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2570/)

**Pull request merged**